src/Images.hs

```diff
@@ -0,0 +1,98 @@
+{-# LANGUAGE ImportQualifiedPost #-}
+{-# LANGUAGE OverloadedStrings #-}
+
+module Images
+  ( createImageTable,
+    getImageFromDB,
+    storeImageInDB,
+    handleGetImage,
+    handleGetAssociatedImageUUIDs,
+  )
+where
+
+import Codec.Picture (DynamicImage, readImage)
+import Codec.Picture.Saving (imageToJpg)
+import Control.Monad.IO.Class (MonadIO (liftIO))
+import Data.ByteString.Lazy qualified as BS (writeFile)
+import Data.UUID (UUID)
+import Data.UUID.V4 (nextRandom)
+import Database.SQLite.Simple (Connection, FromRow (fromRow), Only (Only), ToRow (toRow), execute, execute_, field, query)
+import Database.SQLite.Simple.ToField (ToField (toField))
+import Network.HTTP.Types (status400, status404)
+import Recipe (RecipeID)
+import Text.Read (readMaybe)
+import Util (Occurence (Many, None, One), occurences)
+import Web.Scotty (ActionM, addHeader, json, param, raiseStatus, raw)
+
+createImageTable :: Connection -> IO ()
+createImageTable db = do
+  {-
+    Image:
+      uuid:     UUID of the image
+      title:    Title of the image
+      filePath: Filepath to the image
+      recipeID: ID of the associated recipe
+  -}
+  execute_
+    db
+    "CREATE TABLE IF NOT EXISTS images (\
+    \ uuid VARCHAR(36) PRIMARY KEY NOT NULL,\
+    \ title VARCHAR(40) NOT NULL,\
+    \ filePath VARCHAR(100) NOT NULL,\
+    \ recipeId INTEGER,\
+    \ FOREIGN KEY (recipeId) REFERENCES recipes(recipeId));"
+
+data ImageInfo = ImageInfo
+  { uuid :: UUID,
+    title :: String,
+    filePath :: FilePath,
+    recipeId :: RecipeID
+  }
+
+instance FromRow ImageInfo where
+  fromRow = ImageInfo <$> (read <$> field) <*> field <*> field <*> field
+
+instance ToRow ImageInfo where
+  toRow (ImageInfo imageUUID imageTitle fPath rId) = [toField . show $ imageUUID, toField imageTitle, toField fPath, toField rId]
+
+imageFolder :: FilePath
+imageFolder = "images/"
+
+getImageFromDB :: Connection -> UUID -> IO (Occurence DynamicImage)
+getImageFromDB db imageUUID = occurences <$> (traverse readImageFromDisk =<< query db "SELECT * FROM images WHERE uuid = ?;" (Only . show $ imageUUID))
+  where
+    readImageFromDisk info = either error id <$> (readImage . filePath $ info)
+
+storeImageOnDisk :: FilePath -> DynamicImage -> IO ()
+storeImageOnDisk fPath = BS.writeFile fPath . imageToJpg 50
+
+storeImageInDB :: Connection -> String -> DynamicImage -> RecipeID -> IO UUID
+storeImageInDB db imageTitle image rId = do
+  randomUUID <- nextRandom
+  let fPath = imageFolder ++ show randomUUID ++ ".jpg"
```

Comment:
hier fehlt u.U. ein "/", wenn der Pfad konfigurierbar ist. Gibt es kein Path-Concat in Haskell?

Reply:
Resolved by e75f8bc5613b05785fa0ad2992cda0b4fa7b4aab